cli/tools/script/interpreter.rs

```diff
@@ -0,0 +1,248 @@
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::process::Stdio;
+use std::time::Duration;
+
+use deno_core::anyhow::anyhow;
+use deno_core::anyhow::bail;
+use deno_core::error::AnyError;
+use deno_core::futures;
+use deno_core::futures::future::BoxFuture;
+use deno_core::futures::FutureExt;
+
+use crate::fs_util;
+use crate::tools::script::shell_parser::SequentialItem;
+
+use super::shell_parser::EnvVar;
+use super::shell_parser::ListOperator;
+use super::shell_parser::SequentialList;
+use super::shell_parser::ShellCommand;
+use super::shell_parser::ShellCommandBody;
+
+#[derive(Clone)]
+struct EnvState {
+  pub env_vars: HashMap<String, String>,
+  pub cwd: PathBuf,
+}
+
+impl EnvState {
+  pub fn apply_env_var(&mut self, var: &EnvVar) {
+    if var.value.is_empty() {
+      self.env_vars.remove(&var.name);
+    } else {
+      self.env_vars.insert(var.name.clone(), var.value.clone());
+    }
+  }
+
+  pub fn apply_change(&mut self, change: &EnvChange) {
+    match change {
+      EnvChange::SetEnvVar(var) => self.apply_env_var(var),
+      EnvChange::Cd(new_dir) => {
+        self.cwd = new_dir.clone();
+      }
+    }
+  }
+
+  pub fn apply_changes(&mut self, changes: &[EnvChange]) {
+    for change in changes {
+      self.apply_change(change);
+    }
+  }
+}
+
+enum EnvChange {
+  SetEnvVar(EnvVar),
+  Cd(PathBuf),
+}
+
+pub async fn execute(
+  list: SequentialList,
+  env_vars: HashMap<String, String>,
+  cwd: PathBuf,
+  additional_cli_args: Vec<String>,
+) -> Result<i32, AnyError> {
+  assert!(cwd.is_absolute());
+  let list = append_cli_args(list, additional_cli_args)?;
+  let mut state = EnvState { env_vars, cwd };
+  let mut async_handles = Vec::new();
+  let mut final_exit_code = 0;
+  for item in list.items {
+    match item {
+      SequentialItem::Async(item) => {
+        let state = state.clone();
+        async_handles.push(tokio::task::spawn(async move {
+          let result = execute_command(item, state).await;
+          if let ExecuteCommandResult::Continue(_, _, Some(err)) = result {
+            eprintln!("{}", err);
+          }
+        }));
+      }
+      SequentialItem::Command(item) => {
+        match execute_command(item, state.clone()).await {
```

Comment:
Shouldn't `state` be mutated here instead of being cloned?

Reply:
I don't think lifetimes would allow it. Within `execute_command` it returns a boxed future because `execute_command` is async recursive. So for that reason, I have it returning a list of changes to apply to the state after returning (there's a bug with cd that I've fixed in my local)